firedrake/linear_solver.py

```diff
@@ -116,36 +118,27 @@ def _rhs(self):
 
         u = function.Function(self.trial_space)
         b = cofunction.Cofunction(self.test_space.dual())
-        expr = -action(self.A.a, u)
-        return u, get_assembler(expr).assemble, b
+        expr = - action(self.A.a, u)
+        return u, get_assembler(expr, bcs=self.A.bcs, zero_bc_nodes=False, needs_zeroing=False).assemble, b
 
     def _lifted(self, b):
```

Comment:
This is too hard to understand.

Reply:
I now spitted this into three separate cached properties (`_ulift`, `_blift`, and `_update_rhs`), and added some comments explaining how the BC lifting can be done in a single call to `assemble`.